### PR TITLE
feat(isthmus): improved Calcite support for Substrait Aggregate rels 

### PR DESCRIPTION
**Pull request by vbarua**
Substrait Aggregates that contain expressions that are not field
references and/or grouping keys that are not in input order require
extra processing to be converted to Calcite Aggregates successfully AND
correctly

### CODE REVIEW

**Comment by vbarua** [comment]
Fixes https://github.com/substrait-io/substrait-java/issues/155

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java [hunk 1]
```diff
@@ -440,48 +479,75 @@ public AggregateFunctionInvocation count(Rel input, int field) {
...
+    return min(fieldReference(input, field));
+  }
+
+  public AggregateFunctionInvocation min(Expression expr) {
     return singleArgumentArithmeticAggregate(
-        input, field, "min", TypeCreator.asNullable(inputType));
+        expr,
+        "min",
+        // min output is always nullable
+        TypeCreator.asNullable(expr.getType()));
```
Comment (vbarua) [inline core/src/main/java/io/substrait/dsl/SubstraitBuilder.java:490]: Added arithmetic aggregate functions that take expression directly.

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java [hunk 2]
```diff
@@ -57,35 +57,63 @@ public SubstraitBuilder(SimpleExtension.ExtensionCollection extensions) {
...
+        aggsFn.andThen(
+            functionInvocations ->
+                functionInvocations.stream().map(this::measure).collect(Collectors.toList()));
+    return aggregateM(groupingFn, measuresFn, input);
+  }
+
+  public Aggregate aggregateM(
+      Function<Rel, Aggregate.Grouping> groupingFn,
+      Function<Rel, List<Aggregate.Measure>> measuresFn,
       Rel input) {
```
Comment (vbarua) [inline core/src/main/java/io/substrait/dsl/SubstraitBuilder.java:82]: I tried to add this as an overload of `aggregate` like:
```java
  public Aggregate aggregate(
      Function<Rel, Aggregate.Grouping> groupingFn,
      Function<Rel, List<Aggregate.Measure>> measuresFn,
      Rel input) {
```
but it results in clashes as both methods have the same erasure.

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java [hunk 3]
```diff
@@ -494,6 +560,48 @@ private AggregateFunctionInvocation singleArgumentArithmeticAggregate(
...
+    return arithmeticFunction("substract", left, right);
+  }
+
+  public Expression.ScalarFunctionInvocation multiply(Expression left, Expression right) {
+    return arithmeticFunction("multiply", left, right);
+  }
+
+  public Expression.ScalarFunctionInvocation divide(Expression left, Expression right) {
+    return arithmeticFunction("divide", left, right);
+  }
```
Comment (vbarua) [inline core/src/main/java/io/substrait/dsl/SubstraitBuilder.java:587]: I needed to nest some functions within aggregate function calls for testing purposes, so I added some easy ones.

File: isthmus/src/main/java/io/substrait/isthmus/AggregateValidator.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+ * org.apache.calcite.rel.core.Aggregate} rels
+ *
+ * <p>The code in this class can:
+ *
+ * <ul>
+ *   <li>Check for these cases
+ *   <li>Rewrite the Substrait {@link Aggregate} such that it can be converted to Calcite
+ * </ul>
+ */
+public class AggregateValidator {
```
Comment (vbarua) [inline isthmus/src/main/java/io/substrait/isthmus/AggregateValidator.java:27]: I'm open to suggestions for a better name for this.

File: isthmus/src/main/java/io/substrait/isthmus/SubstraitRelVisitor.java [hunk 5]
```diff
@@ -271,6 +271,7 @@ Aggregate.Measure fromAggCall(RelNode input, Type.Struct inputType, AggregateCal
     if (call.filterArg != -1) {
       builder.preMeasureFilter(FieldReference.newRootStructReference(call.filterArg, inputType));
     }
+    // TODO: handle the collation on the AggregateCall
```
Comment (vbarua) [inline isthmus/src/main/java/io/substrait/isthmus/SubstraitRelVisitor.java:274]: Filed an issues for this: https://github.com/substrait-io/substrait-java/issues/215

**Review by patientstreetlight — APPROVED** [review]
Mostly had minor style questions/comments, but overall I think this looks good, and very welcome.

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java [hunk 6]
```diff
@@ -57,35 +57,63 @@ public SubstraitBuilder(SimpleExtension.ExtensionCollection extensions) {
...
+
+  public Aggregate.Measure measure(AggregateFunctionInvocation aggFn, Expression preMeasureFilter) {
+    return Aggregate.Measure.builder().function(aggFn).preMeasureFilter(preMeasureFilter).build();
+  }
+
   public Aggregate aggregate(
       Function<Rel, Aggregate.Grouping> groupingFn,
-      Function<Rel, List<AggregateFunctionInvocation>> measuresFn,
+      Function<Rel, List<AggregateFunctionInvocation>> aggsFn,
```
[the remark this answers is not in the document]
Comment (vbarua) [reply core/src/main/java/io/substrait/dsl/SubstraitBuilder.java:70]: The SubstraitBuilder is intended to help construct plan trees in code. It's more akin to a DSQL than a builder (I probably should have called it something else in hindsight).

 So for example when building:
```java
              b.aggregate(
                  input -> b.grouping(input, 0, 2),
                  input -> List.of(b.count(input, 0)),
                  <inputRel>));
```
`input` is set to `inputRel`, and is eventually passed to a FieldReference constructor which can use the input to set it's type.

Without functions, with a signature like:
```java
  public Aggregate aggregate(
      Aggregate.Grouping groupingFn,
      List<AggregateFunctionInvocation> measuresFn
      List<AggregateFunctionInvocation>
)
```

You could either do something like
```java
var inputRel = <inputRel> // subtree
var agg = 
              b.aggregate(
                  b.grouping(inputRel, 0, 2),
                  List.of(b.count(inputRel, 0)),
                  <inputRel>));
```
and build your relation tree piecemeal and still get the nice field reference utilities or

```java
var inputRel = <inputRel>
var agg = 
              b.aggregate(
                  b.grouping(
                     b.fieldReference(0, <Type Of Input Field 0>), 
                     b.fieldReference(2, <Type Of Input Field 2>)),
                  List.of(b.count(b.fieldReference(0, <Type Of Input Field 0>))),
                  <inputRel>));
```
and create the field references yourself, type and all.

TLDR: Using functions like this let's us construct field references more ergonomically. Building trees piecemeal would also work as a design.

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java [hunk 7]
```diff
@@ -57,35 +57,63 @@ public SubstraitBuilder(SimpleExtension.ExtensionCollection extensions) {
...
+      Rel input) {
+    Function<Rel, List<Aggregate.Measure>> measuresFn =
+        aggsFn.andThen(
+            functionInvocations ->
+                functionInvocations.stream().map(this::measure).collect(Collectors.toList()));
+    return aggregateM(groupingFn, measuresFn, input);
+  }
+
+  public Aggregate aggregateM(
```
[the remark this answers is not in the document]
Comment (vbarua) [reply core/src/main/java/io/substrait/dsl/SubstraitBuilder.java]: Thinking about this some more, I removed the the `aggregateM` methods and updated the `aggregate` methods to consume `Aggregate.Measure` instead of `AggregateFunctionInvocation`. I then updated the aggregate functions methods to return `Aggregate.Measure` as well.

Because the `aggregate` functions now consume `Aggregate.Measure`, we can now layer in additional utils to tweak the measure as needed like:
```java
            input ->
                List.of(
                    withPreMeasureFilter(
                        b.sum(input, 0), b.equal(b.fieldReference(input, 1), b.i32(42)))),
```

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java [hunk 8]
```diff
@@ -426,6 +459,12 @@ public Aggregate.Grouping grouping(Rel input, int... indexes) {
     return Aggregate.Grouping.builder().addAllExpressions(columns).build();
   }
 
+  public Aggregate.Grouping grouping(Expression... expressions) {
+    List<Expression> exprs =
+        Arrays.stream(expressions).collect(java.util.stream.Collectors.toList());
+    return Aggregate.Grouping.builder().addAllExpressions(exprs).build();
```
[the remark this answers is not in the document]
Comment (vbarua) [reply core/src/main/java/io/substrait/dsl/SubstraitBuilder.java:442]: I don't think this works because `addAllExpressions` requires an `Iterable<? extends Expression> elements`, not an   `Iterator<Expression>`.

That being said, there is an `addExpressions(Expression... elements)` methods that can be used directly 😅 .

File: isthmus/src/main/java/io/substrait/isthmus/AggregateValidator.java (hunk 4 above)
Comment (vbarua) [reply isthmus/src/main/java/io/substrait/isthmus/AggregateValidator.java:27]: This inspired me to use `PreCalciteAggregateValidator` and `PreCalciteAggregateTransformer`.

It's very similar to `AggregatePreCalciteTransformer` but with extra emphasis on the `PreCalcite` part of it.

**Comment by vbarua** [comment]
To briefly summarise what this PR is about. Calcite has a more restrictive definition of aggregate relations than Substrait does, which means that not all Substrait aggregates can be converted to Calcite aggregates. The primary restriction is that Calcite aggregates can only take field references in various places that Substrait allows full expressions.

Luckily, this is easy to address by simply moving those expressions to a project before the aggregate, and then updating the aggregate to use references.

The various comments in the `PreCalciteAggregateValdidator` try and document these various restrictions in more detail.

**Comment by vbarua** [comment]
Addresses https://github.com/substrait-io/substrait-java/issues/155

File: core/src/main/java/io/substrait/dsl/SubstraitBuilder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (patientstreetlight) [reply core/src/main/java/io/substrait/dsl/SubstraitBuilder.java]: fwiw `() -> Arrays.stream(expressions).iterator()` _is_ an `Iterable`, since it is a lambda which, when called, produces an `Iterator`.

File: isthmus/src/main/java/io/substrait/isthmus/PreCalciteAggregateValidator.java
Comment (vibhatha) [inline isthmus/src/main/java/io/substrait/isthmus/PreCalciteAggregateValidator.java:1]: @vbarua I am not very fluent with the Calcite component, so I will leave someone else to verify this part.

File: isthmus/src/test/java/io/substrait/isthmus/ComplexAggregateTest.java
Comment (vibhatha) [inline isthmus/src/test/java/io/substrait/isthmus/ComplexAggregateTest.java:1]: General comment: I see the expected output is hand-calculated (meaning determined by ourselves), is there a way to use Calcite to get the optimal and validate with it? This is an out of scope area for me, but just curious if we can do things differently.

**Comment by vibhatha** [comment]
@vbarua seems like there is a conflict.

**Review by vibhatha — APPROVED** [review]
@vbarua Thanks for working on this. The overall change LGTM! 
Though I would appreciate someone with better Calcite knowledge take a look.

File: isthmus/src/test/java/io/substrait/isthmus/ComplexAggregateTest.java
Comment (vbarua) [reply isthmus/src/test/java/io/substrait/isthmus/ComplexAggregateTest.java:1]: The transformation code that I'm testing here is purely going from Substrait -> Substrait. My intent with these tests was to demonstrate that the `transformToValidCalciteAggregate(aggregate)` produces equivalent _Substrait_ plans, no Calcite required.

Calcite only becomes relevant in the validation code when we call
```java
    // Substrait POJO -> Calcite
    new SubstraitToCalcite(EXTENSION_COLLECTION, typeFactory).convert(pojo);
```
which tries to convert the given input to a Calcite plan. One thing that's hard to show here is that if you comment out
```java
    if (!PreCalciteAggregateValidator.isValidCalciteAggregate(aggregate)) {
      aggregate =
          PreCalciteAggregateValidator.PreCalciteAggregateTransformer
              .transformToValidCalciteAggregate(aggregate);
    }
```
in the `SubstraitRelNodeConverter#visit(Aggregate aggregate)` method all of these fail.

> is there a way to use Calcite to get the optimal and validate with it?

Something that's kind of cursed about this is that Calcite already does something like this, but _only_ when you give it SQL input. When it converts from its SqlNode representation to the RelNode representation it applies similar transforms to this.

I don't think copying exactly what Calcite does is a good idea here, because I'd rather perform straightforward (if verbose) transformation here that may not necessarily be the _best_ but is easier to reason about. This also happens _before_ any actual optimizations a user would perform with Calcite, as this transformation is just to allow us to convert the Substrait plan into a Calcite representation.

Does this answer what you were asking?

**Comment by vbarua** [comment]
I'm going to merge these changes. They will only take effect for substrait plans that would previously have resulted in exceptions, so they shouldn't be surprising anyone.
